# 9차시 실습 — 머신러닝 예측 모델 만들기 (정답)

`합격여부`는 1 = 합격, -1 = 불합격입니다(1~8차시와 같은 표기).

## 1단계. 데이터 읽고 빈칸(결측값) 정리하기

In [1]:
import pandas as pd

df = pd.read_csv("../../data/weekly/week09/week09_pass_fail_train.csv", encoding="utf-8-sig")
print(df.shape)

(750, 9)


In [2]:
print(df.isna().sum())

공정명          0
설비번호         0
온도_섭씨        1
압력_Pa        3
가스유량_slm     3
두께_nm        0
진공도_mTorr    0
습도_pct       8
합격여부         0
dtype: int64


In [3]:
df = df.dropna()
print(df.shape)

(735, 9)


## 2단계. Feature(문제)와 Label(정답) 나누기

In [4]:
feature_cols = ["온도_섭씨", "압력_Pa", "가스유량_slm", "두께_nm", "진공도_mTorr", "습도_pct"]

X = df[feature_cols]   # Feature(문제): 공정 조건 6개
print(X.shape)

(735, 6)


In [5]:
y = df["합격여부"]       # Label(정답): 1 = 합격, -1 = 불합격
print(y.value_counts())

합격여부
 1    582
-1    153
Name: count, dtype: int64


## 3단계. 연습문제와 모의고사로 나누기(train/test)

In [6]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)
print(X_train.shape, X_test.shape)

(588, 6) (147, 6)


## 4단계. 결정트리 모델 학습시키기

In [7]:
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier(max_depth=4, random_state=42)
model.fit(X_train, y_train)

UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 2950: illegal multibyte sequence

UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 2950: illegal multibyte sequence

DecisionTreeClassifier(max_depth=4, random_state=42)

## 5단계. 모의고사 채점하기 — 정확도와 "찍기" 비교

In [8]:
from sklearn.metrics import accuracy_score

pred = model.predict(X_test)
accuracy = accuracy_score(y_test, pred)
print(f"모델 정확도: {accuracy:.1%}")

모델 정확도: 80.3%


### 무조건 '합격'이라고 찍으면?

In [9]:
baseline = (y_test == 1).mean()
print(f"무조건 '합격'이라고 찍었을 때 정확도: {baseline:.1%}")

무조건 '합격'이라고 찍었을 때 정확도: 78.9%


## 6단계. 혼동행렬로 네 칸 나눠 보기

In [10]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, pred, labels=[1, -1])
print(cm)

[[109   7]
 [ 22   9]]


## 7단계(종합). 모델이 많이 참고한 열 확인하기 — 원인 후보

In [11]:
for name, importance in zip(feature_cols, model.feature_importances_):
    print(f"{name}: {importance:.1%}")

온도_섭씨: 48.8%
압력_Pa: 16.7%
가스유량_slm: 9.2%
두께_nm: 4.2%
진공도_mTorr: 21.1%
습도_pct: 0.0%


### 도전 문제 — 질문을 8번까지 늘리면? (빠른 학습자용)

In [12]:
deep_model = DecisionTreeClassifier(max_depth=8, random_state=42)
deep_model.fit(X_train, y_train)

UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 2950: illegal multibyte sequence

UnicodeDecodeError: 'cp949' codec can't decode byte 0xe2 in position 2950: illegal multibyte sequence

DecisionTreeClassifier(max_depth=8, random_state=42)

In [13]:
print(f"max_depth=4 연습문제(train): {accuracy_score(y_train, model.predict(X_train)):.1%}")
print(f"max_depth=4 모의고사(test): {accuracy_score(y_test, model.predict(X_test)):.1%}")
print(f"max_depth=8 연습문제(train): {accuracy_score(y_train, deep_model.predict(X_train)):.1%}")
print(f"max_depth=8 모의고사(test): {accuracy_score(y_test, deep_model.predict(X_test)):.1%}")

max_depth=4 연습문제(train): 84.7%
max_depth=4 모의고사(test): 80.3%
max_depth=8 연습문제(train): 94.2%
max_depth=8 모의고사(test): 78.9%


## 8단계. 결과를 문장으로 정리하기
모범 문장: 모의고사 147건에서 모델 정확도는 80.3%로, 무조건 합격이라고 찍은 78.9%보다 조금 높았다. 실제 불합격 31건 중 9건을 잡았고 22건을 놓쳤다. 모델이 가장 많이 참고한 열은 온도_섭씨(48.8%)이지만, 이것은 원인이 아니라 원인 후보다.

## 9단계. AI에게 질문하며 더 알아보기
질문 예시는 강의 "AI로 한 걸음 더" 절에 있습니다. 답은 학생마다 다릅니다.

## 10단계. AI에게 코드 생성 요청하고 직접 실행해보기
강의의 "AI가 만들어줄 수 있는 코드 예시"입니다.

In [14]:
ok_pass = cm[0][0]       # 합격을 합격으로 맞힘
false_alarm = cm[0][1]   # 합격인데 불합격으로 의심
missed = cm[1][0]        # 불합격인데 합격으로 놓침
caught = cm[1][1]        # 불합격을 불합격으로 잡아냄

print(f"실제 합격 {ok_pass + false_alarm}건 중 {ok_pass}건은 맞혔고, {false_alarm}건은 불합격으로 잘못 의심했습니다.")
print(f"실제 불합격 {missed + caught}건 중 {caught}건은 잡아냈고, {missed}건은 합격으로 놓쳤습니다.")

실제 합격 116건 중 109건은 맞혔고, 7건은 불합격으로 잘못 의심했습니다.
실제 불합격 31건 중 9건은 잡아냈고, 22건은 합격으로 놓쳤습니다.


## 11단계. 연습문제 — 진공도 하나만 넣어 보기

In [15]:
vac_model = DecisionTreeClassifier(max_depth=4, random_state=42)
vac_model.fit(X_train[["진공도_mTorr"]], y_train)
vac_pred = vac_model.predict(X_test[["진공도_mTorr"]])

print(f"진공도 하나만 넣은 모델 정확도: {accuracy_score(y_test, vac_pred):.1%}")
print(confusion_matrix(y_test, vac_pred, labels=[1, -1]))

진공도 하나만 넣은 모델 정확도: 76.9%
[[110   6]
 [ 28   3]]


## 12단계. AI로 재미있는 미니 프로그램 만들기 — 합격 예측 미니 데모
강의 예시 코드입니다.

In [16]:
sample = pd.DataFrame([{
    "온도_섭씨": 301.5, "압력_Pa": 1010.0, "가스유량_slm": 52.0,
    "두께_nm": 100.5, "진공도_mTorr": 5.0, "습도_pct": 45.0,
}])  # 실제로는 input()으로 각 값을 하나씩 입력받아도 좋다.

result = model.predict(sample[feature_cols])[0]
if result == 1:   # 1 = 합격, -1 = 불합격
    print("합격일 것 같아요!")
else:
    print("불합격 위험이 있어요. 공정 조건을 다시 확인해 보세요.")

합격일 것 같아요!
